### DISCUSSION TITLE
Swapping edges in directed graph

### DISCUSSION THREAD

**Discussion by Ashwin-op** [networkx/networkx#5222] [Q&A] [answered]
For swapping edges in an undirected graph there is a method `double_edge_swap` in networkx, is there a similar one for swapping edges in directed graphs? I searched the documentation but couldn't find one

**Comment by rossbar** [comment]
It's not 100% clear to me what you mean by "swapping edges in directed graphs". Do you mean it in the same sense as `double_edge_swap`, i.e. an operation that randomly swaps edges between a set of four nodes while preserving the degree of each node?

**Reply by Ashwin-op** [reply]
Yes something similar to that

**Reply by rossbar** [reply]
In that case, I believe @dschult 's response is the answer, so I will mark it as such!

**Comment by dschult** [comment, marked as answer]
The short answer is "no" we don't have a function to perform directed edge swaps.  

It is more difficult to maintain the degree distributions for directed graphs than for undirected graphs. And it is harder to ensure that the edge swaps are sampling the set of all DiGraphs with that degree distribution -- and are they sampling in a useful way?
   
Here is [a reference for directed edge swapping](https://epubs.siam.org/doi/abs/10.1137/16M1087175) that discusses the difficulties of maintaining the degree distribution when the graph is directed. It states that you would need a triangle reversal swap in order to sample graphs uniformly via edge swapping.

> Directed graphs: Sampling directed graphs using edge swap Markov chains introduces new
subtleties that are not present when sampling undirected graphs. Most importantly, a directed
graph has two separate degree sequences, the in-degree sequence and out-degree sequence, and
one may wish to fix either or both of these. The two sequences are coupled because the sum of
the graph’s in-degree must equal the sum of its out-degrees. Furthermore, in order for a graph
of directed graphs to be connected under edge swaps, a directed triangle reversal swap is needed,
Figure 6(a), which reverses the direction of a three-edge cycle [75, 117, 78]. Sampling both stub-labeled and vertex-labeled directed graphs builds on a similar theoretical framework as undirected graphs [25, 26, 27].

> Broadly speaking, as richer network models are considered the sensible value of uniform distributions as statistical null models decreases. Developing an appropriate null model for richer
networks, where directed graphs are one example, requires carefully considering and modeling a
hypothesized generative processes. For example, a directed version of a citation network should
roughly obey causality constraints (cycles would indicate past papers citing future papers), and the
statistical properties of such a network might be best captured by comparing it to the output of
generative model that explicitly accounts for publication date.
